curriculum/challenges/english/07-scientific-computing-with-python/learn-encapsulation-by-building-a-projectile-trajectory-calculator/66475d40dd42fb614f14fd03.md

```diff
@@ -0,0 +1,212 @@
+
+    def __init__(self, speed, height, angle):
+        self.__speed = speed
+        self.__height = height
+        self.__angle = math.radians(angle)
+        
+    def __str__(self):
+        return f'''
+Projectile details:
+speed: {self.__speed} m/s
+height: {self.__height} m
+angle: {round(math.degrees(self.__angle))}°
+displacement: {round(self.__calculate_displacement(), 1)} m
+'''
+
+    def __calculate_displacement(self):
+        horizontal_component = self.__speed * math.cos(self.__angle)
+        vertical_component = self.__speed * math.sin(self.__angle)
+        squared_component = vertical_component**2
+        gh_component = 2 * GRAVITATIONAL_ACCELERATION * self.__height
+        sqrt_component = math.sqrt(squared_component + gh_component)
+        
+        displacement = horizontal_component * (vertical_component + sqrt_component) / GRAVITATIONAL_ACCELERATION
+        return displacement
+        
+    def __calculate_y_coordinate(self, x):
+        height_component = self.__height
+        angle_component = math.tan(self.__angle) * x
+        acceleration_component = GRAVITATIONAL_ACCELERATION * x ** 2 / (
+                2 * self.__speed ** 2 * math.cos(self.__angle) ** 2)
+        y_coordinate = height_component + angle_component - acceleration_component
+
+        return y_coordinate
+    
+    def calculate_all_coordinates(self):
+        return [
+            (x, self.__calculate_y_coordinate(x))
+            for x in range(math.ceil(self.__calculate_displacement()))
+        ]
+
+    @property
+    def height(self):
+        return self.__height
+
+    @property
+    def angle(self):
+        return round(math.degrees(self.__angle))
+
+    @property
+    def speed(self):
+        return self.__speed
+
+    @height.setter
+    def height(self, n):
+        self.__height = n
+
+    @angle.setter
+    def angle(self, n):
+        self.__angle = math.radians(n)
+
+    @speed.setter
+    def speed(self, s):
+       self.__speed = s
+    
+    def __repr__(self):
+        return f'{self.__class__}({self.speed}, {self.height}, {self.angle})'
+
+class Graph:
+    __slots__ = ('__coordinates')
+
+    def __init__(self, coord):
+        self.__coordinates = coord
+
+    def __repr__(self):
+        return f"Graph({self.__coordinates})"
+
+    def create_coordinates_table(self):
+        table = '\n  x      y\n'
+        for x, y in self.__coordinates:
+            table += f'{x:>3}{round(y, 2):>7.2f}\n'
+
+        return table
+
+--fcc-editable-region--
+    def create_trajectory(self):
+        
+        rounded_coords = [(round(x), round(y)) for x,y in self.__coordinates]
+        
+        x_max = max(rounded_coords, key=lambda i: i[0])[0]
+        y_max = max(rounded_coords, key=lambda j: j[1])[1]
+
+        return x_max, y_max
+        
+        
+--fcc-editable-region--
+
+ball = Projectile(10, 3, 45)
+print(ball)
+coordinates = ball.calculate_all_coordinates()
+graph = Graph(coordinates)
+print(graph.create_trajectory())
```

Comment:
This isn't very readable in next steps, which might make it harder to get the correct results.

Maybe another step could be added to change it for couple steps?:
```python
for row in graph.create_trajectory()
    print(row)
```